server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,263 @@
+        logger.trace("syncMetaData() called ");
+        writeLock.lock();
+        try {
+            localDirectory.syncMetaData();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void rename(String source, String dest) throws IOException {
+        logger.trace("rename() called {}, {}", source, dest);
+        writeLock.lock();
+        try {
+            localDirectory.rename(source, dest);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        logger.trace("openInput() called {}", name);
+        writeLock.lock();
+        try {
+            if (Arrays.asList(getRemoteFiles()).contains(name) == false) {
+                // If file has not yet been uploaded to Remote Store, fetch it from the local directory
+                logger.trace("File found in disk");
+                return localDirectory.openInput(name, context);
+            } else {
+                // If file has been uploaded to the Remote Store, fetch it from the Remote Store in blocks via
+                // OnDemandCompositeBlockIndexInput
+                logger.trace("File to be fetched from Remote");
+                return new OnDemandCompositeBlockIndexInput(remoteDirectory, name, localDirectory, fileCache, context);
+            }
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public Lock obtainLock(String name) throws IOException {
+        logger.trace("obtainLock() called {}", name);
+        writeLock.lock();
+        try {
+            return localDirectory.obtainLock(name);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        writeLock.lock();
+        try {
+            localDirectory.close();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public Set<String> getPendingDeletions() throws IOException {
+        readLock.lock();
+        try {
+            return localDirectory.getPendingDeletions();
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    /**
+     * Function to perform operations once files have been uploaded to Remote Store
+     * Once uploaded local files are safe to delete
+     * @param files : files which have been successfully uploaded to Remote Store
+     * @throws IOException
+     */
+    public void afterSyncToRemote(Collection<String> files) throws IOException {
+        logger.trace("afterSyncToRemote called for {}", files);
+        if (remoteDirectory == null) {
+            logger.trace("afterSyncToRemote called even though remote directory is not set");
+            return;
+        }
+        for (String fileName : files) {
+            writeLock.lock();
+            try {
+                localDirectory.deleteFile(fileName);
```

Comment:
Let's add a comment here for context. We probably don't want to delete a file as soon as its uploaded to remote. We are doing that right now just to test the block fetch code path as the Cache pinning & eviction hasn't be introduced yet.

Reply:
+1 on this one.